### PR TITLE
Find ER based on workspace

### PR DESCRIPTION
**Pull request by anandjakhaniya**

### CODE REVIEW

File: backend/src/main/java/ai/verta/modeldb/experimentRun/FutureExperimentRunDAO.java [hunk 1]
```diff
@@ -532,30 +613,89 @@ public FutureExperimentRunDAO(
...
+          getAllowedProjects(ModelDBActionEnum.ModelDBServiceActions.READ)
+              .thenCompose(
+                  projectIds -> {
+                    if (projectIds.isEmpty()) {
+                      // TODO: This is not works, need to update it.
+                      /*return InternalFuture.failedStage(new PermissionDeniedException(
+                      "Accessible projects not found in workspace: "
+                              + request.getWorkspaceName()));*/
+                    }
+                    return InternalFuture.completedInternalFuture(projectIds);
```
Comment (anandjakhaniya) [inline backend/src/main/java/ai/verta/modeldb/experimentRun/FutureExperimentRunDAO.java]: @conradoverta , @hmacdonald-verta, I am throwing error from here but it is not works and request is stuck somewhere so i don't understand what i did wrong and where i have writing wrong InternalFuture. Could you check and suggest me here please??

File: backend/src/main/java/ai/verta/modeldb/experimentRun/FutureExperimentRunDAO.java [hunk 2]
```diff
@@ -532,30 +613,89 @@ public FutureExperimentRunDAO(
...
+    if (request.getWorkspaceName().isEmpty()) {
+      queryFilterContextList.add(
+          getAllowedProjects(ModelDBActionEnum.ModelDBServiceActions.READ)
+              .thenCompose(
+                  projectIds -> {
+                    if (projectIds.isEmpty()) {
+                      // TODO: This is not works, need to update it.
+                      /*return InternalFuture.failedStage(new PermissionDeniedException(
+                      "Accessible projects not found in workspace: "
+                              + request.getWorkspaceName()));*/
```
Comment (conradoverta) [inline backend/src/main/java/ai/verta/modeldb/experimentRun/FutureExperimentRunDAO.java]: Why doesn't it work?

File: backend/src/main/java/ai/verta/modeldb/experimentRun/FutureExperimentRunDAO.java (hunk 1 above)
Comment (conradoverta) [reply backend/src/main/java/ai/verta/modeldb/experimentRun/FutureExperimentRunDAO.java]: I don't unserstand what is not working and how you know it's not working

File: backend/src/main/java/ai/verta/modeldb/experimentRun/FutureExperimentRunDAO.java [hunk 3]
```diff
@@ -532,30 +613,89 @@ public FutureExperimentRunDAO(
...
+                              workspace.getId(),
+                              ModelDBResourceEnum.ModelDBServiceResourceTypes.PROJECT)
+                          .thenCompose(
+                              getResourcesResponseItems -> {
+                                if (getResourcesResponseItems.isEmpty()) {
+                                  // TODO: This is not works, need to update it.
+                                  /*return InternalFuture.failedStage(
+                                  new PermissionDeniedException(
+                                      "Accessible project not found in workspace: "
+                                          + request.getWorkspaceName()));*/
```
Comment (conradoverta) [inline backend/src/main/java/ai/verta/modeldb/experimentRun/FutureExperimentRunDAO.java]: Same question

File: backend/src/main/java/ai/verta/modeldb/experimentRun/FutureExperimentRunDAO.java [hunk 4]
```diff
@@ -744,9 +884,7 @@ public FutureExperimentRunDAO(
                 executor);
 
     final var futureCount =
-        InternalFuture.sequence(
-                Arrays.asList(futureLocalContext, futurePredicatesContext, futureProjectIds),
-                executor)
+        InternalFuture.sequence(queryFilterContextList, executor)
```
Comment (conradoverta) [inline backend/src/main/java/ai/verta/modeldb/experimentRun/FutureExperimentRunDAO.java]: This doesn't require any sorting context.

File: backend/src/main/java/ai/verta/modeldb/experimentRun/FutureExperimentRunDAO.java (hunk 1 above)
Comment (anandjakhaniya) [reply backend/src/main/java/ai/verta/modeldb/experimentRun/FutureExperimentRunDAO.java]: I have try `return InternalFuture.failedStage` but MDB stuck after that statement and if i am trying to throw `throw new PermissionDeniedException` then it is also stuck there. i don't what is happened. Could you suggest me something for it please?

File: backend/src/main/java/ai/verta/modeldb/experimentRun/FutureExperimentRunDAO.java (hunk 2 above)
Comment (anandjakhaniya) [reply backend/src/main/java/ai/verta/modeldb/experimentRun/FutureExperimentRunDAO.java]: I have try `return InternalFuture.failedStage` but MDB stuck after that statement and if i am trying to throw `throw new PermissionDeniedException` then it is also stuck there. i don't what is happened. Could you suggest me something for it please?

File: backend/src/main/java/ai/verta/modeldb/experimentRun/FutureExperimentRunDAO.java (hunk 4 above)
Comment (anandjakhaniya) [reply backend/src/main/java/ai/verta/modeldb/experimentRun/FutureExperimentRunDAO.java]: it is not for sorting, i have some condition based InternalFuture so i have added list to keep all future and then added list in `InternalFuture.sequence`

Comment (conradoverta) [reply backend/src/main/java/ai/verta/modeldb/experimentRun/FutureExperimentRunDAO.java]: You are adding the sorting context. You don't need it.

https://github.com/VertaAI/docs/blob/main/engineering-practices/api-implementation/performance.md#avoid-layers-of-abstraction

File: backend/src/main/java/ai/verta/modeldb/experimentRun/FutureExperimentRunDAO.java [hunk 5]
```diff
@@ -536,30 +617,89 @@ public FutureExperimentRunDAO(
...
+    // futurePredicatesContext
+    queryFilterContextList.add(
+        predicatesHandler.processPredicates(request.getPredicatesList(), executor));
+
+    // futureSortingContext
+    queryFilterContextList.add(
+        sortingHandler.processSort(request.getSortKey(), request.getAscending()));
+
+    // futureProjectIds based on workspace
+    if (request.getWorkspaceName().isEmpty()) {
```
Comment (conradoverta) [inline backend/src/main/java/ai/verta/modeldb/experimentRun/FutureExperimentRunDAO.java]: This whole logic is way too big to be inlined in this method. Extract it to a different method.

File: backend/src/main/java/ai/verta/modeldb/experimentRun/FutureExperimentRunDAO.java (hunk 2 above)
Comment (anandjakhaniya) [reply backend/src/main/java/ai/verta/modeldb/experimentRun/FutureExperimentRunDAO.java]: Fixed in PR https://github.com/VertaAI/modeldb/pull/2242

File: backend/src/main/java/ai/verta/modeldb/experimentRun/FutureExperimentRunDAO.java (hunk 3 above)
Comment (anandjakhaniya) [reply backend/src/main/java/ai/verta/modeldb/experimentRun/FutureExperimentRunDAO.java]: Fixed in PR https://github.com/VertaAI/modeldb/pull/2242#pullrequestreview-648893239

File: backend/src/main/java/ai/verta/modeldb/experimentRun/FutureExperimentRunDAO.java (hunk 5 above)
Comment (anandjakhaniya) [reply backend/src/main/java/ai/verta/modeldb/experimentRun/FutureExperimentRunDAO.java]: extracted in separate method

File: backend/src/main/java/ai/verta/modeldb/experimentRun/FutureExperimentRunDAO.java (hunk 4 above)
Comment (anandjakhaniya) [reply backend/src/main/java/ai/verta/modeldb/experimentRun/FutureExperimentRunDAO.java]: Yes i realize that and removed it now

File: backend/src/main/java/ai/verta/modeldb/experimentRun/FutureExperimentRunDAO.java [hunk 6]
```diff
@@ -348,6 +401,34 @@ public FutureExperimentRunDAO(
...
+    ResourceType resourceType =
+        ResourceType.newBuilder()
+            .setModeldbServiceResourceType(modelDBServiceResourceTypes)
+            .build();
+    Resources.Builder resources =
+        Resources.newBuilder()
+            .setResourceType(resourceType)
+            .setService(ServiceEnum.Service.MODELDB_SERVICE);
+
+    if (resourceIds != null && !resourceIds.isEmpty()) {
```
Comment (conradoverta) [inline backend/src/main/java/ai/verta/modeldb/experimentRun/FutureExperimentRunDAO.java]: Make it optional for clarity on the fact that it's optional.

File: backend/src/main/java/ai/verta/modeldb/experimentRun/FutureExperimentRunDAO.java [hunk 7]
```diff
@@ -806,4 +935,36 @@ public FutureExperimentRunDAO(
...
         executor);
   }
+
+  private InternalFuture<List<String>> getAccessibleProjectIdsBasedOnWorkspace(
+      String workspaceName, String projectId) {
+    if (workspaceName.isEmpty()) {
+      return getAllowedProjects(ModelDBActionEnum.ModelDBServiceActions.READ);
+    } else {
+      var requestProjectIds = new ArrayList<String>();
+      if (!projectId.isEmpty()) {
```
Comment (conradoverta) [inline backend/src/main/java/ai/verta/modeldb/experimentRun/FutureExperimentRunDAO.java]: Optional

**Review by conradoverta — APPROVED** [review]
Two optional asks. LGTM otherwise

**Pull request merged**